### PR TITLE
Advanced streamconn usage

### PR DESCRIPTION
**Pull request by shlomikushchi**
Add the ability to "pause/resume" the websocket connection.
also added advanced ws example code for:
* pause / resume connection
* change subscriptions of existing connection
* ws disconnections handler (make sure we reconnect)
* Different subscriptions usage

### CODE REVIEW

File: alpaca_trade_api/polygon/streamconn.py [hunk 1]
```diff
@@ -190,12 +190,15 @@ def run(self, initial_channels=[]):
...
     async def close(self):
         '''Close any open connections'''
-        if self._consume_task:
-            self._consume_task.cancel()
+        await self.cancel_task()
         if self._ws is not None:
             await self._ws.close()
         self._ws = None
 
+    async def cancel_task(self):
```
Comment (gnvk) [inline alpaca_trade_api/polygon/streamconn.py:198]: [minor] should be private: `_cancel_task()`

File: alpaca_trade_api/stream2.py [hunk 2]
```diff
@@ -140,12 +141,15 @@ async def unsubscribe(self, channels):
...
 
     async def close(self):
-        if self._consume_task:
-            self._consume_task.cancel()
+        await self.cancel_task()
         if self._ws:
             await self._ws.close()
             self._ws = None
 
+    async def cancel_task(self):
```
Comment (gnvk) [inline alpaca_trade_api/stream2.py:149]: [minor] should be private: `_cancel_task()`

File: alpaca_trade_api/stream2.py [hunk 3]
```diff
@@ -227,6 +231,7 @@ def __init__(
         self._data_stream = _data_stream
         self._debug = debug
         self._raw_data = raw_data
+        self.stop_stream_queue = queue.Queue()
```
Comment (gnvk) [inline alpaca_trade_api/stream2.py:234]: [minor] should be private: `_stop_stream_queue`

File: alpaca_trade_api/stream2.py [hunk 4]
```diff
@@ -370,6 +378,19 @@ async def close(self, renew):
                                            self._oauth,
                                            raw_data=self._raw_data)
 
+    async def signal_stop_ws(self):
```
Comment (gnvk) [inline alpaca_trade_api/stream2.py:381]: [minor] I would simply call this `stop()` or `stop_ws()`

File: alpaca_trade_api/stream2.py [hunk 5]
```diff
@@ -370,6 +378,19 @@ async def close(self, renew):
...
+        """
+        Signal the ws connections to stop.
+        We set a signal in a Queue and then call cancel_task
```
Comment (gnvk) [inline alpaca_trade_api/stream2.py:384]: [minor] this is an implementation detail that I wouldn't include in the docstring.

File: alpaca_trade_api/stream2.py (hunk 3 above)
Comment (shlomikushchi) [reply alpaca_trade_api/stream2.py:234]: done

File: alpaca_trade_api/stream2.py (hunk 4 above)
Comment (shlomikushchi) [reply alpaca_trade_api/stream2.py:381]: done

File: alpaca_trade_api/stream2.py (hunk 5 above)
Comment (shlomikushchi) [reply alpaca_trade_api/stream2.py:384]: done. removed internal implementation from doc string

File: alpaca_trade_api/polygon/streamconn.py (hunk 1 above)
Comment (shlomikushchi) [reply alpaca_trade_api/polygon/streamconn.py:198]: it is used outside of the class so it shouldn't be private

File: alpaca_trade_api/stream2.py (hunk 2 above)
Comment (shlomikushchi) [reply alpaca_trade_api/stream2.py:149]: same. used out side of the class.

**Review by gnvk — APPROVED** [review]
LGTM

**Pull request merged**